sensor/common/scannerclient/v4_grpc_client.go

```diff
@@ -0,0 +1,157 @@
+	scannerV1 "github.com/stackrox/scanner/generated/scanner/api/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+var _ Client = (*V4GRPCClient)(nil)
+var languageComponents = map[string]bool{
+	"Go":       true,
+	"Maven":    true,
+	"PiPI":     true,
+	"RubyGems": true,
+}
+
+// V4GRPCClient represents a client implementation using the v4 gRPC protocol.
+type V4GRPCClient struct {
+	indexerClient scannerV4.IndexerClient
+	matcherClient scannerV4.MatcherClient
+	conn          *grpc.ClientConn
+}
+
+func (v V4GRPCClient) Dial(endpoint string) (Client, error) {
+	if endpoint == "" {
+		return nil, errors.New("Invalid Scanner endpoint (empty)")
+	}
+
+	endpoint = strings.TrimPrefix(endpoint, "https://")
+	if strings.Contains(endpoint, "://") {
+		return nil, errors.Errorf("ScannerV4 endpoint has unsupported scheme: %s", endpoint)
+	}
+
+	tlsConfig, err := clientconn.TLSConfig(mtls.ScannerSubject, clientconn.TLSConfigOptions{
+		UseClientCert: clientconn.MustUseClientCert,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize ScannerV4 TLS config")
+	}
+
+	// This is non-blocking. If we ever want this to block,
+	// then add the grpc.WithBlock() DialOption.
+	conn, err := grpc.Dial(endpoint, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to dial Scanner")
+	}
+
+	log.Infof("Dialing ScannerV4 at %s", endpoint)
+
+	return &V4GRPCClient{
+		indexerClient: scannerV4.NewIndexerClient(conn),
+		matcherClient: scannerV4.NewMatcherClient(conn),
+		conn:          conn,
+	}, nil
+}
+
+func (v V4GRPCClient) GetImageAnalysis(ctx context.Context, image *storage.Image, cfg *types.Config) (*scannerV1.GetImageComponentsResponse, error) {
+	name := image.GetName().GetFullName()
+	indexReport, err := v.indexerClient.CreateIndexReport(ctx, &scannerV4.CreateIndexReportRequest{
+		HashId:               "",
+		ResourceLocator:      nil,
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     nil,
+		XXX_sizecache:        0,
+	}, grpc.WaitForReady(true))
+
+	if err != nil {
+		log.Debugf("Unable to get image components from local Scanner for image %s: %v", name, err)
+		return nil, errors.Wrap(err, "getting image components from scanner")
+	}
+
+	log.Debugf("Received image indexing report from local Scanner for image %s", name)
+
+	//convert indexReport to scannerV4.CreateIndexReportResponse
+	resp, err := convertIndexReportToV1GetImageComponentsResponse(*indexReport, image)
+	//return resp or return indexReport directly?
+	if err != nil {
+		log.Debugf("Failed to convert indexer report to image components from local Scanner for image %s: %v", name, err)
+		return nil, errors.Wrap(err, "converting indexer report from scanner")
+	}
+
+	return resp, nil
+}
+
+func convertIndexReportToV1GetImageComponentsResponse(indexReport scannerV4.IndexReport, image *storage.Image) (*scannerV1.GetImageComponentsResponse, error) {
```

Comment:
Have you made sure that ImageComponents has all the necessary components for the Central's Scanner V4 Indexer and Matcher to be able to match vulnerabilities? 

Additionally, we need to make decisions about how the Matcher will retrieve the IndexReport to match vulnerabilities. I'm curious to know if we plan to store it in the database or somewhere else for the Matcher to access. See https://issues.redhat.com/browse/ROX-16919

Reply:
The second point of view makes sense to me. If we are going to save index report to DB for matcher to fetch them later, do we still have to worry about the first question?